File: Lab2/src/student_code.py
import common
def astar_search(map):
	found = False
	# PUT YOUR CODE HERE
	# access the map using "map[y][x]"
	# y between 0 and common.constants.MAP_HEIGHT-1
	# x between 0 and common.constants.MAP_WIDTH-1
	start = (0,0)
	end = (0,0)
    	for i in range(0, common.constants.MAP_HEIGHT):
        	for j in range(0, common.constants.MAP_WIDTH):
        	    if(map[i][j] == 2):
                	start = (i,j)
        	    if(map[i][j] == 3):
   	           	end = (i,j)
		#	print(end,start)
	Queue = AstarQueue()
	distFromStart = {}
    	parent = {}
    	parent[start] = None
    	f_initial = manhattan_distance(start, end)
	distFromStart[start] = 0
    	start_y, start_x = start
	current = start
	Queue.push(f_initial, start)
    	while(Queue.length() > 0):
        	y,x = Queue.pop()
       		if map[y][x] == 3:
			map[y][x] = 5
			while(parent[current] != None):
				y,x = current
				map[y][x] = 5
				current = parent[current]
			map[start_y][start_x] = 5
			found =  True
			break
		else:
        	 	map[y][x] = 4
		 	if(y >= 0 and y <common.constants.MAP_HEIGHT and x+1 >= 0 and x+1 < common.constants.MAP_WIDTH and (map[y][x+1] == 0 or map[y][x+1] == 3)):
				current = (y,x+1)
				distFromStart[current] = distFromStart[(y, x)] + 1
				f1 = distFromStart[current] + manhattan_distance(current, end)
				Queue.push(f1, current)
				parent[current] = (y, x)
				
			
			if(y >= 0 and y+1 < common.constants.MAP_HEIGHT and x >= 0 and x < common.constants.MAP_WIDTH and (map[y+1][x] == 0 or map[y+1][x] == 3)):
				current = (y+1,x)
				distFromStart[current] = distFromStart[(y, x)] + 1
				f2 = distFromStart[current] + manhattan_distance(current, end)
				Queue.push(f2,current)
				parent[current] = (y, x)
			
			
			if(y >= 0 and y < common.constants.MAP_HEIGHT and x-1 >= 0 and x-1 < common.constants.MAP_WIDTH and (map[y][x-1] == 0 or map[y][x-1] == 3)):
				current = (y,x-1)
				distFromStart[current] = distFromStart[(y, x)] + 1
				f3 = distFromStart[current] + manhattan_distance(current, end)
				Queue.push(f3,current)
				parent[current] = (y, x)
				
			if(y-1 >= 0 and y-1 < common.constants.MAP_HEIGHT and x >= 0 and x < common.constants.MAP_WIDTH and (map[y-1][x] == 0 or map[y-1][x] == 3)):
				current = (y-1,x)
				distFromStart[current] = distFromStart[(y, x)] + 1
				f4 = distFromStart[current] + manhattan_distance(current, end)
				Queue.push(f4,current)
				parent[current] = (y, x)

	return found

def manhattan_distance(start, end):
	y_start, x_start = start
	y_end, x_end = end
	a = (abs(y_start - y_end) + abs(x_start - x_end))
	return(a)
class AstarQueue(object):
	def __init__(self):
		self.Queue = []
	
	def push(self, new_f, new_point):
		new_y, new_x = new_point
		i = 0
		if not self.Queue:
			self.Queue.append((new_f, new_point))
			return
		else:
			for f, point in self.Queue:
				y, x = point

				if f > new_f:
					i += 1
					
					
				elif f == new_f:

					if x < new_x:
						self.Queue.insert(i, (new_f, new_point))	
						return
					elif x == new_x:

						if y < new_y:
							self.Queue.insert(i, (new_f, new_point))
							return
					else:
						i += 1
					
					

				else: 
					self.Queue.insert(i, (new_f, new_point))
					return
			i += 1
			self.Queue.append((new_f, new_point))
	def length(self):
		return(len(self.Queue))
	def pop(self):
		f, point = self.Queue.pop()
		return point

File: Lab1/src/student_code.c
int queue[MAP_WIDTH*MAP_HEIGHT][2];
int path_x[MAP_HEIGHT][MAP_WIDTH];
int path_y[MAP_HEIGHT][MAP_WIDTH];
bool dfs_recursive( int **map, int y, int x)
{
      bool flag,path;
//	printf("%d %d ", y,x);
      if(y<0 || x<0 || y>= MAP_HEIGHT || x>= MAP_WIDTH || map[y][x] == 1 || map[y][x] == 4 )
          flag = false;
      else
      {
          if(map[y][x] == 3)
          {
            map[y][x] = 5;
            flag = true;
          }
	else
	{

          map[y][x] = 4;
          if(dfs_recursive(map, y, x + 1) || dfs_recursive(map, y + 1, x) || dfs_recursive(map, y, x - 1) || dfs_recursive(map, y - 1, x))  
	{
            map[y][x] = 5;
            flag = true;
          }
          else
            flag = false;
        }
      }
        return flag;
}
bool df_search(int **map)
{
        bool found = false;
        int x,y,i,j = 0;
        // PUT YOUR CODE HERE
        // access the map using "map[y][x]"
        // y between 0 and MAP_HEIGHT-1
        // x between 0 and MAP_WIDTH-1
        for( i=0; i<MAP_HEIGHT; i++)
        {       
        	 for( j = 0; j<MAP_WIDTH; j++)
                 {       
   			 if(map[i][j] == 2)
                         {
				 y = i; x = j;                                       
                          }                                                                                                                                                                                                   }
	}
     found = dfs_recursive(map, y, x);
	printf("%d ", found);      
      	return found;
}                                                                                                                                  
		
void enqueue(int y, int x, int &front, int &rear)
{	
	if(front == -1 || rear == -1)
		front = rear = 0;
	else
		rear = rear + 1;
	queue[rear][0] = y;
	queue[rear][1] = x;

}
void dequeue(int &y, int &x, int &rear, int &front)
{	
	
//	parent = front;
	y = queue[front][0];
	x = queue[front][1];
	if(front == rear)
		front = rear = -1;
	else
		front = front + 1;

}
	
bool bf_search(int **map)
{
	bool found = false;
	int x,y,i,j = 0;
	int start_x,start_y;
	// PUT YOUR CODE HERE
	// access the map using "map[y][x]"
	// y between 0 and MAP_HEIGHT-1
	// x between 0 and MAP_WIDTH-1
	 for( i=0; i<MAP_HEIGHT; i++)
        {
                for( j = 0; j<MAP_WIDTH; j++)
                {
                        if(map[i][j] == 2)
                        {
                                y = i; x = j;
				start_y = i; start_x = j;			
                        }
                }
        }
	int front, rear;
	queue[0][0] = y;
	queue[0][1] = x;
	front = 0; rear = 0; i = 0;
	int new_x,new_y;
	while(front != -1 && rear!= -1 && front <= rear && rear < MAP_HEIGHT*MAP_WIDTH)
	{	
		int parent_id = rear;
		dequeue(y, x, rear, front);
		if(map[y][x] == 3)
		{
			found = true;	
			new_x = x; new_y = y;
			while(!(start_y ==  y ))
			{	
				y = new_y; x = new_x;
				map[y][x] = 5;	
				new_y = path_y[y][x];
				new_x = path_x[y][x];
			} 
			break;
		}
		
		else
		{
			map[y][x] = 4;
			if(!(y<0 || x+1<0 || y>= MAP_HEIGHT || x+1>=MAP_WIDTH || map[y][x+1] == 1 || map[y][x+1] == 4))
			{
				enqueue(y, x+1, front, rear);
				path_y[y][x+1] = y;
				path_x[y][x+1] = x;			
			}		
			if(!(y+1<0 || x<0 || y+1>= MAP_HEIGHT || x>=MAP_WIDTH || map[y+1][x] == 1 || map[y+1][x] == 4))
			{
				enqueue(y+1, x, front, rear);
				path_y[y+1][x] = y;  
                                path_x[y+1][x] = x;
			
			}
			if(!(y<0 || x-1<0 || y>= MAP_HEIGHT || x-1>=MAP_WIDTH || map[y][x-1] == 1 || map[y][x-1] == 4))
			{
				 enqueue(y, x-1, front, rear);
				path_y[y][x-1] = y;  
                                path_x[y][x-1] = x;	
			}
			if(!(y-1<0 || x<0 || y-1>= MAP_HEIGHT || x>=MAP_WIDTH || map[y-1][x] == 1 || map[y-1][x] == 4))
		        {
			         enqueue(y-1, x, front, rear);
				path_y[y-1][x] = y;  
                                path_x[y-1][x] = x;
			}	
		}
		
	}
	
	
				
	return found;
}
